modules/examples/sources/threaded-diskq-source/threaded-diskq-source.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2018 Balabit
+ * Copyright (c) 2018 László Várady <laszlo.varady@balabit.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "threaded-diskq-source.h"
+#include "logthrsource/logthrfetcherdrv.h"
+#include "logqueue-disk.h"
+#include "logqueue-disk-reliable.h"
+#include "logqueue-disk-non-reliable.h"
+#include "logmsg/logmsg.h"
+#include "messages.h"
+
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+struct ThreadedDiskqSourceDriver
+{
+  LogThreadedFetcherDriver super;
+  DiskQueueOptions diskq_options;
+  LogQueue *queue;
+
+  struct stat diskq_file_stat;
+  gboolean waiting_for_file_change;
+
+  gchar *filename;
+};
+
+static gboolean
+_load_queue(ThreadedDiskqSourceDriver *self)
+{
+  self->diskq_options.read_only = TRUE;
+  self->diskq_options.reliable = FALSE;
+
+  FILE *file = fopen(self->filename, "rb");
+
+  if (!file)
+    {
+      msg_error("Error opening diskq file", evt_tag_str("file", self->filename));
+      return FALSE;
+    }
+
+  gchar file_signature[5] = {0};
+  if (fread(file_signature, 4, 1, file) == 0)
+    {
+      msg_error("Error reading diskq file signature", evt_tag_str("file", self->filename));
+      fclose(file);
+      return FALSE;
+    }
+
+  fclose(file);
+
+  if (strcmp(file_signature, "SLRQ") == 0)
+    self->diskq_options.reliable = TRUE;
+
+  if (self->diskq_options.reliable)
+    {
+      self->diskq_options.disk_buf_size = 128;
+      self->diskq_options.mem_buf_size = 1024 * 1024;
+      self->queue = log_queue_disk_reliable_new(&self->diskq_options, NULL);
+    }
+  else
+    {
+      self->diskq_options.disk_buf_size = 1;
+      self->diskq_options.mem_buf_size = 128;
+      self->diskq_options.qout_size = 128;
+      self->queue = log_queue_disk_non_reliable_new(&self->diskq_options, NULL);
+    }
+
```

Comment:
this auto-detection and creation of the LogQueue instance should probably be delegated to the logqueue module
somehow.

I know this is only an example plugin but could be a very useful one indeed.

Reply:
I agree, I just didn't want to change existing code in this PR, but I'll do it in another one.